### PR TITLE
Added Alchemy: Dominaria set. Implemented Vinesoul Spider, Slimefoot, Thallid Transplant.

### PR DESCRIPTION
**Pull request by karapuzz14**
Added Vinesoul Spider card and its set "Alchemy: Dominaria". Also I've added function of putting random card into graveyard from your library based on "seek card" realization. It's my first commit so I'm waiting for review and possible changes.

### CODE REVIEW

**Comment by karapuzz14** [comment]
I've also implemented Slimefoot, Thallid Transplant card and reprints of existing cards. Cards should work correctly but there's one problem which I don't know how to solve. Card Viewer shows that 11/23 cards are implemented when there's 39 existing cards in set. I've written all of them into mtg-cards-data.txt but Card Viewer doesn't see them.

**Comment by JayDi85** [comment]
> that 11/23 cards are implemented when there're 39 existing cards in set 

It’s ok, card viewer used a max card number of implemented cards. Txt files used for fast card class generation (if you run python script from Utils folder).

File: Mage.Client/src/main/java/org/mage/plugins/card/dl/sources/GathererSets.java [hunk 1]
```diff
@@ -104,7 +104,7 @@ private CheckResult(String ACode, ExpansionSet ASet, boolean AHaveCommon, boolea
             "UNF", "BRO", "BRC", "BOT", "30A", "J22", "SCD", "DMR", "ONE", "ONC",
             "MOM", "MOC", "MUL", "MAT", "LTR", "CMM", "WOE", "WHO", "RVR", "WOT",
             "WOC", "SPG", "LCI", "LCC", "REX", "PIP", "MKM", "MKC", "CLU", "OTJ",
-            "OTC", "OTP", "BIG", "MH3", "ACR", "BLB"
+            "OTC", "OTP", "BIG", "MH3", "ACR", "BLB", "YDMU"
```
Comment (xenohedron) [inline Mage.Client/src/main/java/org/mage/plugins/card/dl/sources/GathererSets.java:107]: please keep in order, looks like should be after DMC, before 40K

File: Mage.Sets/src/mage/cards/v/VinesoulSpider.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        this.subtype.add(SubType.SPIDER);
+
+        this.power = new MageInt(3);
+        this.toughness = new MageInt(2);
+
+        this.addAbility(ReachAbility.getInstance());
+
+        // At the beginning of your end step, put a random land card from your library into your graveyard.
+        this.addAbility(new BeginningOfEndStepTriggeredAbility(
+                new mage.cards.v.VinesoulSpiderEffect(), TargetController.YOU, false
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/v/VinesoulSpider.java:34]: don't include the `mage.cards.v`, the effect is already in the namespace (many instances to fix in this file)

File: Mage/src/main/java/mage/players/PlayerImpl.java [hunk 3]
```diff
@@ -2881,6 +2881,22 @@ public boolean seekCard(FilterCard filter, Ability source, Game game) {
         return true;
     }
 
+    @Override
+    public boolean putRandomCardInGraveyard(FilterCard filter, Ability source, Game game) {
```
Comment (xenohedron) [inline Mage/src/main/java/mage/players/PlayerImpl.java:2885]: I don't agree with putting this as inner method in `Player` classes. It's only used in two cards as far as I can tell? What's wrong with putting the code only in the custom card class where it is used?

File: Mage.Sets/src/mage/sets/AlchemyDominaria.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        return instance;
+    }
+
+    private AlchemyDominaria() {
+        super("Alchemy: Dominaria", "YDMU", ExpansionSet.buildDate(2022, 10, 5), SetType.MAGIC_ARENA);
+        this.blockName = "Alchemy";
+        this.hasBoosters = false;
+        this.hasBasicLands = false;
+
+        cards.add(new SetCardInfo("Vinesoul Spider", 18, Rarity.UNCOMMON, mage.cards.v.VinesoulSpider.class));
```
Comment (xenohedron) [inline Mage.Sets/src/mage/sets/AlchemyDominaria.java:24]: typically all cards, new and reprint, are sorted together alphabetically

**Review by xenohedron — CHANGES_REQUESTED** [review]

File: Mage/src/main/java/mage/players/PlayerImpl.java (hunk 3 above)
Comment (Susucre) [reply Mage/src/main/java/mage/players/PlayerImpl.java:2885]: Agreeing with the sentiment.
Either in the custom Effect, or in a shared Effect that could handle all those: https://scryfall.com/search?q=o%3A%22put+a+random%22&unique=cards&as=grid&order=name

File: Mage/src/main/java/mage/abilities/effects/common/PutRandomCardFromLibraryIntoGraveyardEffect.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    public PutRandomCardFromLibraryIntoGraveyardEffect copy() {
+        return new PutRandomCardFromLibraryIntoGraveyardEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        Permanent sourcePermanent = game.getPermanentOrLKIBattlefield(source.getSourceId());
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/effects/common/PutRandomCardFromLibraryIntoGraveyardEffect.java:44]: sourcePermanent isn't used anywhere?

**Review by xenohedron — COMMENTED** [review]
rest looks good

File: Mage/src/main/java/mage/abilities/effects/common/PutRandomCardFromLibraryIntoGraveyardEffect.java (hunk 5 above)
Comment (karapuzz14) [reply Mage/src/main/java/mage/abilities/effects/common/PutRandomCardFromLibraryIntoGraveyardEffect.java:44]: > sourcePermanent isn't used anywhere?

It's assumed that sourcePermanent is used for confirmation of source permanent existence (at line 45).

Comment (xenohedron) [reply Mage/src/main/java/mage/abilities/effects/common/PutRandomCardFromLibraryIntoGraveyardEffect.java:44]: Not necessary. The permanent doesn't have to still be on the battlefield for the effect to resolve, which is why it's checking LKI, but the LKI isn't relevant to anything about the effect either.

Comment (karapuzz14) [reply Mage/src/main/java/mage/abilities/effects/common/PutRandomCardFromLibraryIntoGraveyardEffect.java]: Okay. Then I should remove it.

**Review by xenohedron — APPROVED** [review]
thanks for the contribution

**Pull request merged**